corehq/apps/domain/tests/test_domain_name_generation.py

```diff
@@ -0,0 +1,64 @@
+from __future__ import print_function, unicode_literals
+
+from django.test import SimpleTestCase, TestCase
+
+from corehq.apps.domain.models import Domain
+from corehq.apps.domain.exceptions import NameUnavailableException
+
+
+class DomainNameAvailabilityTest(SimpleTestCase):
+    def test_conflict(self):
+        self.assertEquals(Domain._get_next_available_name("fandango", []), "fandango-1")
+
+    def test_availability(self):
+        name = "abc"
+        similar = []
+
+        self.assertEquals(Domain._get_next_available_name(name, similar), "abc-1")
+
+        similar.append("abc-1")
+        self.assertEquals(Domain._get_next_available_name(name, similar), "abc-2")
+
+        # Don't bother filling in gaps
+        similar.append("abc-9")
+        self.assertEquals(Domain._get_next_available_name(name, similar), "abc-10")
```

Comment:
I think the "one user enters their year of birth and everyone else gets a huge number" one is the main thing, but I won't push too hard if you don't think it's worth it